toolkit/tools/srpmpacker/srpmpacker.go

```diff
@@ -193,26 +263,119 @@ func createAllSRPMs(specsDir, distTag, buildDir, outDir string, workers int, nes
+	} else {
+		for _, specName := range packList {
+			var specFile []string
+
+			specSearch := filepath.Join(specsDir, fmt.Sprintf("**/%s.spec", specName))
+			specFile, err = filepath.Glob(specSearch)
+
+			// If a SPEC is in the pack list, it must be packed.
+			if err != nil {
+				return
+			}
+			if len(specFile) != 1 {
+				err = fmt.Errorf("unexpected number of matches (%d) for spec file (%s)", len(specFile), specName)
+				return
+			}
+
+			specFiles = append(specFiles, specFile[0])
+		}
+	}
+
+	return
+}
+
+// createChroot creates a chroot to pack SRPMs inside of.
+func createChroot(workerTar, buildDir, outDir, specsDir string) (chroot *safechroot.Chroot, newBuilDir, newOutDir, newSpecsDir string, err error) {
+	const (
+		chrootName       = "srpmpacker_chroot"
+		existingDir      = false
+		leaveFilesOnDisk = false
+
+		outMountPoint    = "/output"
+		specsMountPoint  = "/specs"
+		buildDirInChroot = "/build"
+	)
+
+	extraMountPoints := []*safechroot.MountPoint{
+		safechroot.NewMountPoint(outDir, outMountPoint, "", safechroot.BindMountPointFlags, ""),
+		safechroot.NewMountPoint(specsDir, specsMountPoint, "", safechroot.BindMountPointFlags, ""),
+	}
+
+	extraDirectories := []string{
+		buildDirInChroot,
+	}
+
+	newBuilDir = buildDirInChroot
+	newOutDir = outMountPoint
+	newSpecsDir = specsMountPoint
+
+	chrootDir := filepath.Join(buildDir, chrootName)
+	chroot = safechroot.NewChroot(chrootDir, existingDir)
+
+	err = chroot.Initialize(workerTar, extraDirectories, extraMountPoints)
+	if err != nil {
+		return
+	}
+
+	// If this is container build then the bind mounts will not have been created.
+	// Copy in all of the SPECs so they can be packed.
+	if !buildpipeline.IsRegularBuild() {
+		specsInChroot := filepath.Join(chroot.RootDir(), newSpecsDir)
+		err = directory.CopyContents(specsDir, specsInChroot)
+		if err != nil {
+			closeErr := chroot.Close(leaveFilesOnDisk)
+			if closeErr != nil {
+				logger.Log.Errorf("Failed to close chroot, err: %s", err)
+			}
+			return
+		}
+	}
+
+	// Networking support is needed to download sources.
+	files := []safechroot.FileToCopy{
+		{Src: "/etc/resolv.conf", Dest: "/etc/resolv.conf"},
+	}
+
+	err = chroot.AddFiles(files...)
+	if err != nil {
+		chroot.Close(leaveFilesOnDisk)
+	}
+
+	return
+}
+
 // calculateSPECsToRepack will check which SPECs should be packed.
 // If the resulting SRPM does not exist, or is older than a modification to
 // one of the files used by the SPEC then it is repacked.
 func calculateSPECsToRepack(specFiles []string, distTag, outDir string, nestedSourcesDir, repackAll bool, workers int) (states []*specState, err error) {
-	logger.Log.Infof("Calculating SPECs to repack")
+	var wg sync.WaitGroup
 
-	allSpecFiles := make(chan string, len(specFiles))
-	specResults := make(chan *specState, len(specFiles))
+	requests := make(chan string, len(specFiles))
+	results := make(chan *specState, len(specFiles))
+	cancel := make(chan struct{})
```

Comment:
>cancel := make(chan struct{}) [](start = 1, length = 29)

Nit: a channel seems a bit like an overkill for this purpose. Wouldn't a boolean work just as well with passing its pointer to other threads?

Reply:
This concurrency pattern of using a cancellation channel for thread synchronization and leveraging go's select statement is the recommend design pattern in golang: https://blog.golang.org/pipelines

Reply:
Interesting, fine by me, thanks.

---
In reply to: [501147980](https://github.com/microsoft/CBL-Mariner/pull/141#discussion_r501147980) [](ancestors = 501147980)